reactive/ovn_chassis_charm_handlers.py

```diff
@@ -127,3 +127,24 @@ def configure_nrpe():
     """Handle config-changed for NRPE options."""
     with charm.provide_charm_instance() as charm_instance:
         charm_instance.render_nrpe()
+
+
+@reactive.when_none('charm.paused', 'is-update-status-hook')
+@reactive.when(OVN_CHASSIS_ENABLE_HANDLERS_FLAG,
+               'ovsdb-subordinate.available',
+               'certificates.available')
+def provide_chassis_certificates_to_principal():
+    ovsdb_subordinate = reactive.endpoint_from_flag(
+        'ovsdb-subordinate.available')
+    try:
+        with charm.provide_charm_instance() as charm_instance:
+            with open(charm_instance.options.ovn_ca_cert, 'r') as ovn_ca_cert:
+                with open(charm_instance.options.ovn_cert, 'r') as ovn_cert:
+                    with open(charm_instance.options.ovn_key, 'r') as ovn_key:
```

Comment:
Could I suggest two alternatives to the "march to the right" with the with statements.  Either:

```python
with (charm.provide_charm_instance() as charm_instance,
        open(...) as ovn_ca_cert,
        open(...) as ovn_cert,
        open(...) as ovn_key):
    do_stuff()
```

or:

```python
from contextlib import ExitStack

with ExitStack() as es:
    charm_instance = es.enter_context(charm.provide_charm_instance())
    ovn_ca_cert = es.enter_context(open(...))
    ovn_cert = es.enter_context(open(...))
    do_stuff()
```

Reply:
Ah, fantastic, I was not aware of the possibility. Thx for the tip, much appreciated @ajkavanagh !

Reply:
On the surface the former suggestion appears to work in Py3.9 only (why can't we have nice things?), I'll try the second approach.